File: frontend/src/lib/server/caption-formatter.ts
import { config } from "./config";
import type { TranscriptSegment } from "./gemini";

export interface CaptionBlock {
  id: string;
  sequence: number;
  startMs: number;
  endMs: number;
  originalText: string;
  cleanedText?: string | null;
  speaker: string | null;
  minConfidence: number;
}

const maxLineLength = () => config.maxCaptionLineLength;
const maxLines = () => config.maxCaptionLines;
const minDurationMs = () => config.minCaptionDurationMs;
const maxDurationMs = () => config.maxCaptionDurationMs;

export function segmentsToCaptions(
  segments: TranscriptSegment[],
  speakerMap?: Record<string, string> | null
): CaptionBlock[] {
  const captions: CaptionBlock[] = [];
  let sequence = 0;

  for (const seg of segments) {
    const text = seg.text;
    const speakerLabel = seg.speaker
      ? (speakerMap?.[seg.speaker] ?? seg.speaker)
      : null;

    const segMinConf = seg.words.length
      ? Math.min(...seg.words.map((w) => w.confidence))
      : 0.92;

    const chunks = splitIntoCaptionChunks(text);
    if (!chunks.length) continue;

    if (chunks.length === 1) {
      sequence++;
      const prefix = speakerLabel ? `>> ${speakerLabel}: ` : "";
      captions.push({
        id: `cap-${sequence}`,
        sequence,
        startMs: seg.startMs,
        endMs: seg.endMs,
        originalText: prefix + chunks[0],
        speaker: seg.speaker,
        minConfidence: segMinConf,
      });
    } else {
      const totalDuration = seg.endMs - seg.startMs;
      const totalChars = chunks.reduce((s, c) => s + c.length, 0);
      let currentStart = seg.startMs;

      for (let i = 0; i < chunks.length; i++) {
        const chunkRatio =
          totalChars > 0 ? chunks[i].length / totalChars : 1 / chunks.length;
        let chunkDuration = Math.round(totalDuration * chunkRatio);
        chunkDuration = Math.max(
          minDurationMs(),
          Math.min(maxDurationMs(), chunkDuration)
        );
        const chunkEnd = Math.min(currentStart + chunkDuration, seg.endMs);

        sequence++;
        const prefix = i === 0 && speakerLabel ? `>> ${speakerLabel}: ` : "";
        captions.push({
          id: `cap-${sequence}`,
          sequence,
          startMs: currentStart,
          endMs: chunkEnd,
          originalText: prefix + chunks[i],
          speaker: seg.speaker,
          minConfidence: segMinConf,
        });
        currentStart = chunkEnd;
      }
    }
  }

  return captions;
}

export function splitIntoCaptionChunks(text: string): string[] {
  if (!text.trim()) return [];

  const words = text.split(/\s+/);
  const chunks: string[] = [];
  let currentLines: string[] = [];
  let currentLine = "";

  for (const word of words) {
    const testLine = currentLine ? `${currentLine} ${word}` : word;

    if (testLine.length <= maxLineLength()) {
      currentLine = testLine;
    } else {
      if (currentLine) currentLines.push(currentLine);
      currentLine = word;

      if (currentLines.length >= maxLines()) {
        chunks.push(currentLines.join("\n"));
        currentLines = [];
      }
    }
  }

  if (currentLine) currentLines.push(currentLine);
  if (currentLines.length) chunks.push(currentLines.join("\n"));

  return chunks;
}

function formatVttTime(ms: number): string {
  const hours = Math.floor(ms / 3_600_000);
  const minutes = Math.floor((ms % 3_600_000) / 60_000);
  const seconds = Math.floor((ms % 60_000) / 1_000);
  const millis = ms % 1_000;
  return `${String(hours).padStart(2, "0")}:${String(minutes).padStart(2, "0")}:${String(seconds).padStart(2, "0")}.${String(millis).padStart(3, "0")}`;
}

function formatSrtTime(ms: number): string {
  const hours = Math.floor(ms / 3_600_000);
  const minutes = Math.floor((ms % 3_600_000) / 60_000);
  const seconds = Math.floor((ms % 60_000) / 1_000);
  const millis = ms % 1_000;
  return `${String(hours).padStart(2, "0")}:${String(minutes).padStart(2, "0")}:${String(seconds).padStart(2, "0")},${String(millis).padStart(3, "0")}`;
}

export function toVtt(
  captions: Array<Record<string, unknown>>,
  useCleaned = true
): string {
  const lines = ["WEBVTT", ""];
  for (const cap of captions) {
    const text = useCleaned
      ? ((cap.cleaned_text as string) || (cap.original_text as string))
      : (cap.original_text as string);
    lines.push(
      `${formatVttTime(cap.start_ms as number)} --> ${formatVttTime(cap.end_ms as number)}`
    );
    lines.push(text);
    lines.push("");
  }
  return lines.join("\n");
}

export function toSrt(
  captions: Array<Record<string, unknown>>,
  useCleaned = true
): string {
  const lines: string[] = [];
  for (let i = 0; i < captions.length; i++) {
    const cap = captions[i];
    const text = useCleaned
      ? ((cap.cleaned_text as string) || (cap.original_text as string))
      : (cap.original_text as string);
    lines.push(String(i + 1));
    lines.push(
      `${formatSrtTime(cap.start_ms as number)} --> ${formatSrtTime(cap.end_ms as number)}`
    );
    lines.push(text);
    lines.push("");
  }
  return lines.join("\n");
}

export function toTxt(
  captions: Array<Record<string, unknown>>,
  useCleaned = true
): string {
  const lines: string[] = [];
  let currentSpeaker: string | null = null;
  for (const cap of captions) {
    const text = useCleaned
      ? ((cap.cleaned_text as string) || (cap.original_text as string))
      : (cap.original_text as string);
    const speaker = cap.speaker as string | null;
    if (speaker && speaker !== currentSpeaker) {
      if (lines.length) lines.push("");
      lines.push(`[${speaker}]`);
      currentSpeaker = speaker;
    }
    lines.push(text);
  }
  return lines.join("\n");
}

File: backend/app/services/caption_formatter.py
from app.config import get_settings
from app.models.schemas import TranscriptSegment, CaptionBlock


class CaptionFormatter:
    """Generates compliant caption files from transcript segments."""

    def __init__(self):
        settings = get_settings()
        self.max_line_length = settings.max_caption_line_length
        self.max_lines = settings.max_caption_lines
        self.max_reading_speed_wpm = settings.max_reading_speed_wpm
        self.min_duration_ms = settings.min_caption_duration_ms
        self.max_duration_ms = settings.max_caption_duration_ms

    def segments_to_captions(
        self,
        segments: list[TranscriptSegment],
        speaker_map: dict[str, str] | None = None,
    ) -> list[CaptionBlock]:
        """Convert transcript segments into compliant caption blocks."""
        captions: list[CaptionBlock] = []
        sequence = 0

        for seg in segments:
            text = seg.text
            speaker_label = None
            if seg.speaker:
                display_name = (speaker_map or {}).get(seg.speaker, seg.speaker)
                speaker_label = display_name

            seg_min_conf = min((w.confidence for w in seg.words), default=0.92) if seg.words else 0.92
            chunks = self._split_into_caption_chunks(text)

            if not chunks:
                continue

            if len(chunks) == 1:
                sequence += 1
                prefix = f">> {speaker_label}: " if speaker_label else ""
                captions.append(CaptionBlock(
                    id=f"cap-{sequence}",
                    sequence=sequence,
                    start_ms=seg.start_ms,
                    end_ms=seg.end_ms,
                    original_text=prefix + chunks[0],
                    speaker=seg.speaker,
                    min_confidence=seg_min_conf,
                ))
            else:
                total_duration = seg.end_ms - seg.start_ms
                total_chars = sum(len(c) for c in chunks)

                current_start = seg.start_ms
                for i, chunk in enumerate(chunks):
                    chunk_ratio = len(chunk) / total_chars if total_chars > 0 else 1 / len(chunks)
                    chunk_duration = int(total_duration * chunk_ratio)
                    chunk_duration = max(self.min_duration_ms, min(self.max_duration_ms, chunk_duration))
                    chunk_end = min(current_start + chunk_duration, seg.end_ms)

                    sequence += 1
                    prefix = ""
                    if i == 0 and speaker_label:
                        prefix = f">> {speaker_label}: "

                    captions.append(CaptionBlock(
                        id=f"cap-{sequence}",
                        sequence=sequence,
                        start_ms=current_start,
                        end_ms=chunk_end,
                        original_text=prefix + chunk,
                        speaker=seg.speaker,
                        min_confidence=seg_min_conf,
                    ))
                    current_start = chunk_end

        return captions

    def _split_into_caption_chunks(self, text: str) -> list[str]:
        """Split text into chunks of max 2 lines x max_line_length characters."""
        if not text.strip():
            return []

        words = text.split()
        chunks = []
        current_lines: list[str] = []
        current_line = ""

        for word in words:
            test_line = f"{current_line} {word}".strip()

            if len(test_line) <= self.max_line_length:
                current_line = test_line
            else:
                if current_line:
                    current_lines.append(current_line)
                current_line = word

                if len(current_lines) >= self.max_lines:
                    chunks.append("\n".join(current_lines))
                    current_lines = []

        if current_line:
            current_lines.append(current_line)
        if current_lines:
            chunks.append("\n".join(current_lines))

        return chunks

    @staticmethod
    def _format_vtt_time(ms: int) -> str:
        hours = ms // 3_600_000
        minutes = (ms % 3_600_000) // 60_000
        seconds = (ms % 60_000) // 1_000
        millis = ms % 1_000
        return f"{hours:02d}:{minutes:02d}:{seconds:02d}.{millis:03d}"

    @staticmethod
    def _format_srt_time(ms: int) -> str:
        hours = ms // 3_600_000
        minutes = (ms % 3_600_000) // 60_000
        seconds = (ms % 60_000) // 1_000
        millis = ms % 1_000
        return f"{hours:02d}:{minutes:02d}:{seconds:02d},{millis:03d}"

    def to_vtt(self, captions: list[dict], use_cleaned: bool = True) -> str:
        lines = ["WEBVTT", ""]
        for cap in captions:
            text = (cap.get("cleaned_text") or cap["original_text"]) if use_cleaned else cap["original_text"]
            start = self._format_vtt_time(cap["start_ms"])
            end = self._format_vtt_time(cap["end_ms"])
            lines.append(f"{start} --> {end}")
            lines.append(text)
            lines.append("")
        return "\n".join(lines)

    def to_srt(self, captions: list[dict], use_cleaned: bool = True) -> str:
        lines = []
        for i, cap in enumerate(captions, 1):
            text = (cap.get("cleaned_text") or cap["original_text"]) if use_cleaned else cap["original_text"]
            start = self._format_srt_time(cap["start_ms"])
            end = self._format_srt_time(cap["end_ms"])
            lines.append(str(i))
            lines.append(f"{start} --> {end}")
            lines.append(text)
            lines.append("")
        return "\n".join(lines)

    def to_txt(self, captions: list[dict], use_cleaned: bool = True) -> str:
        lines = []
        current_speaker = None
        for cap in captions:
            text = (cap.get("cleaned_text") or cap["original_text"]) if use_cleaned else cap["original_text"]
            speaker = cap.get("speaker")
            if speaker and speaker != current_speaker:
                if lines:
                    lines.append("")
                lines.append(f"[{speaker}]")
                current_speaker = speaker
            lines.append(text)
        return "\n".join(lines)
